File: Game.Api/Models/Items/ItemMod.cs
using Game.Api.Models.Attributes;
using Game.Core;
using ItemModEntity = Game.Abstractions.Entities.ItemMod;

namespace Game.Api.Models.Items
{
    public class ItemMod : IModelFromSource<ItemMod, ItemModEntity>
    {
        public int Id { get; set; }
        public required string Name { get; set; }
        public required string Description { get; set; }
        public EItemModType ItemModTypeId { get; set; }
        public ERarity RarityId { get; set; }
        public required IEnumerable<BattlerAttribute> Attributes { get; set; }
        public required IEnumerable<int> Tags { get; set; }

        public static ItemMod FromSource(ItemModEntity itemMod)
        {
            return new ItemMod
            {
                Id = itemMod.Id,
                Name = itemMod.Name,
                Description = itemMod.Description,
                ItemModTypeId = (EItemModType)itemMod.ItemModTypeId,
                RarityId = (ERarity)itemMod.RarityId,
                Attributes = itemMod.ItemModAttributes.To().Model<BattlerAttribute>(),
                Tags = itemMod.Tags.Select(t => t.Id),
            };
        }
    }
}

File: UI/src/lib/battle/item-mod.ts
import { EItemModType, ERarity, IAppliedModModel, IBattlerAttribute, IItemMod } from '$lib/api';
import { staticData } from '$stores';

// A class (not a plain object literal) so `statify` recognizes it via its own per-field reactivity
// (see `statify.svelte.ts`) instead of falling through to Svelte's native deep-proxy, which would give
// the same underlying mod a different identity depending on which reactive array reads it (#1957).
export class ItemMod implements IItemMod {
	id: number;
	name: string;
	description: string;
	itemModTypeId: EItemModType;
	rarityId: ERarity;
	attributes: IBattlerAttribute[];
	tags: number[];
	designerNotes: string;
	retiredAt?: string;

	itemModSlotId: number;

	constructor(itemModData: IItemMod, itemModSlotId: number) {
		this.id = itemModData.id;
		this.name = itemModData.name;
		this.description = itemModData.description;
		this.itemModTypeId = itemModData.itemModTypeId;
		this.rarityId = itemModData.rarityId;
		this.attributes = itemModData.attributes;
		this.tags = itemModData.tags;
		this.designerNotes = itemModData.designerNotes;
		this.retiredAt = itemModData.retiredAt;
		this.itemModSlotId = itemModSlotId;
	}
}

export const newItemMod = (appliedMod: IAppliedModModel): ItemMod | undefined => {
	// Resolve the mod's static definition by id; a missing/retired id yields no record, so degrade
	// gracefully (mirroring resolveUnlockReward) rather than spreading `undefined` and crashing.
	const itemModData = staticData.itemMods?.[appliedMod.itemModId];
	if (!itemModData) {
		return undefined;
	}
	return new ItemMod(itemModData, appliedMod.itemModSlotId);
};
